src/Components/GEOSldas_GridComp/GEOSlandassim_GridComp/clsm_ensupd_read_obs.F90

```diff
@@ -83,9 +83,439 @@ module clsm_ensupd_read_obs
+             
+             n_read = vsfread( vdata_id(j), tmprealvec, &
+                  N_data_tmp(j), FULL_INTERLACE)
+             
+             CI_Index(k_off+1:k_off+N_data_tmp(j)) = tmprealvec
+             do k=1,N_data_tmp(j)
+             
+                if (CI_Index(k+k_off) <20) then
+                    CI_Index(k+k_off) = -1
+                end if
+             end do
+          case (3)
+             n_read = vsfread( vdata_id(j) ,tmpint2vec, &
+                  N_data_tmp(j), FULL_INTERLACE)
+
+             Cloud_Index(k_off+1:k_off+N_data_tmp(j)) = tmpint2vec
+             
+          case (4)
+             
+             n_read = vsfread(vdata_id(j), tmpint2vec, &
+                  N_data_tmp(j), FULL_INTERLACE)
+             
+             Snow_QA(k_off+1:k_off+N_data_tmp(j)) = tmpint2vec
+ 
+            ! case (5)
+            
+            ! n_read = vsfread(vdata_id(j), tmpint2vec, &
+            !     N_data_tmp(j), FULL_INTERLACE)
+
+            ! lat(k_off+1:k_off+N_data_tmp(j)) = tmpint2vec
+             
+            !case (6)
+            !  n_read = vsfread(vdata_id(j), tmpint2vec, &
+            !      N_data_tmp(j), FULL_INTERLACE)
+
+            !lon(k_off+1:k_off+N_data_tmp(j)) = tmpint2vec
+
+          case default
+             
+          call ldas_abort(LDAS_GENERIC_ERROR, Iam, 'Unknown case')
+          end select
+          
+          if (n_read/=N_data_tmp(j)) &
+           call ldas_abort(LDAS_GENERIC_ERROR, Iam, 'ERROR reading hdf')
+       end do
+       
+
+       ! clean up
+
+       deallocate(tmprealvec)
+       deallocate(tmpint2vec)
+       
+       ! close hdf files
+       
+       status = vsfdtch(vdata_id(j))
+       status = vfend(file_id(j))
+       status = hclose(file_id(j))
+       
+       ! prepare next j
+       k_off = k_off + N_data_tmp(j)
+       
+    end do
+
+    write(HH, '(i2.2)') date_time%hour
+    read(HH, '(I10)') hh_num
+
+    time_index = hh_num/3 +1
+    
+    ! -------------------------------------
+    ! eliminate no-data-values and data that fail initial QC
+    
+    j = 0
+    
+    do i=1,N_data
+       if (logit) write (logunit, *) 'hh_num=', hh_num
+       if (logit) write (logunit, *) 'time_index =', time_index
+
+       if ( (MODIS_SCA(i)>0.)           .and.         &  ! any neg is nodata
+            (CI_Index(i)>20)            .and.         &  ! Ignore obs
+            (Snow_QA(i)<3)              .and.         &
+            (lon(i)>=lon_subtime(time_index)) .and.         &
+            (lon(i)<lon_subtime(time_index+1))              & 
+           ) then     
+          
+          j=j+1
+          
+          MODIS_SCA(j) = MODIS_SCA(i)
+          lon(j)      = lon(i)
+          lat(j)      = lat(i)
+       end if
+       
+    end do
+   
+   if (logit) write (logunit, *) 'assimilation time', HH
+   if (logit) write (logunit, *) 'lat = ', lat
+   if (logit) write (logunit, *) 'lon = ', lon
+   if (logit) write (logunit, *) 'MODIS_SCA=', MODIS_SCA
```

Comment:
Shouldn't these write statements be limited to 1:j (where j is the final element of interest, see N_data=j in the next line)?  As they are, the write statements dump the entire allocated vectors into the file, incl whatever is left after the final N_data.